### PR TITLE
V0.4 Introduce Entry States

### PR DESCRIPTION
**Pull request by ChuaPingChan**
@CS2103JUN2017-T2/developers 
Changes:
1. Added `States` to the `Entry` superclass which can take the enum values `ACTIVE`, `ARCHIVED` and `DELETED`. States can be referred to by importing `Entry` and `Entry.States.*`.
1. Note that entry constructors instantiate entries with active state by default. States can be changed after that with setter methods either invoked directly on entry objects or through Model's API. 
1. The filtered lists binded to the GUI should be able to filter by keywords in names and tags as well as entry's state now. The API to update filtered lists have been changed so that it takes in one more argument of type `Entry.State`.
1. All statements affected by the change to the API to update filtered lists have been corrected so that this PR merges with the main `V0.4` branch without upsetting Travis. Most of the modified calls to `updateFiltered*List()` are in `Logic` @kevinLamKB 

Meanwhile, I'm adding API to `ModelManager` that allows outside components to manipulate entry states inside the `EntryBook`, these are not included in this PR yet.

Part of #111 
Part of #175

### CODE REVIEW

**Comment by ChuaPingChan** [comment]
@CS2103JUN2017-T2/developers 
Hmm I realised this branch actually branched out and merges many other previous branches into it before, so the `Files changed` reflect too many past changes from other branches, making this PR code hard to review. Sorry for that.

**Review by mattheuslee — APPROVED** [review]
Wow that's a lot of changes. I'm gonna take it if all our tests are still okay then there should be no issues haha.

**Comment by ChuaPingChan** [comment]
@mattheuslee yes I made sure Travis passes, so all the tests up to now should be fine haha.

**Comment by kevinLamKB** [comment]
wait how come its merging to master?

**Comment by ChuaPingChan** [comment]
@kevinLamKB Ops, sorry that's a mistake.

**Comment by kevinLamKB** [comment]
> All statements affected by the change to the API to update filtered lists have been corrected so that this PR merges seamlessly with the main V0.4 branch without upsetting Travis. Most of the modified calls to updateFiltered*List() are in Logic @kevinLamKB

👌 should be easy enough to fix, i'll do that on the V0.4 branch after u merge, easier to change with all the error messages from eclipse

**Comment by ChuaPingChan** [comment]
@CS2103JUN2017-T2/developers 
I used the wrong PR base branch `master` instead of `V0.4`. Now that it's corrected, the `Files changed` are more comprehensible already.

File: src/main/java/seedu/multitasky/logic/commands/DeleteByFindCommand.java [hunk 1]
```diff
@@ -19,13 +20,17 @@
...
-            + "Use " + COMMAND_WORD + " [" + String.join(" | ", CliSyntax.PREFIX_EVENT.toString(),
-            CliSyntax.PREFIX_DEADLINE.toString(), CliSyntax.PREFIX_FLOATINGTASK.toString()) + "]"
-            + " INDEX to specify which entry to delete.";
+                                                          + "Use " + COMMAND_WORD + " ["
+                                                          + String.join(" | ",
+                                                                        CliSyntax.PREFIX_EVENT.toString(),
+                                                                        CliSyntax.PREFIX_DEADLINE.toString(),
+                                                                        CliSyntax.PREFIX_FLOATINGTASK.toString())
+                                                          + "]"
+                                                          + " INDEX to specify which entry to delete.";
```
Comment (ChuaPingChan) [inline src/main/java/seedu/multitasky/logic/commands/DeleteByFindCommand.java:30]: @kevinLamKB
Sorry my Eclipse screw this up. Normally changes of this kind are good but not in this case when the strings are long. I'll repair them after dinner. (Same for line 23 in `EditByFindCommand.java`, etc).

Comment (kevinLamKB) [reply src/main/java/seedu/multitasky/logic/commands/DeleteByFindCommand.java:30]: haha, its alright. just leave it as it is. i got the neat copy on my branch still, i can just use my version when i merge into v0.4

**Review by kevinLamKB — APPROVED** [review]

**Comment by mattheuslee** [comment]
@ChuaPingChan Is this ready to merge?

**Pull request merged**